sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@intel.com>
+//
+
+#include <sound/soc.h>
+#include <sound/sof/ipc4/header.h>
+#include <uapi/sound/sof/header.h>
+#include "sof-priv.h"
+#include "ipc4-priv.h"
+#include "sof-client.h"
+#include "sof-client-probes.h"
+#include "ipc4-topology.h"
+
+enum sof_ipc4_dma_type {
+	SOF_IPC4_DMA_HDA_HOST_OUTPUT = 0,
+	SOF_IPC4_DMA_HDA_HOST_INPUT = 1,
+	SOF_IPC4_DMA_HDA_LINK_OUTPUT = 8,
+	SOF_IPC4_DMA_HDA_LINK_INPUT = 9,
+	SOF_IPC4_DMA_DMIC_LINK_INPUT = 11,
+	SOF_IPC4_DMA_I2S_LINK_OUTPUT = 12,
+	SOF_IPC4_DMA_I2S_LINK_INPUT = 13,
+};
+
+union sof_ipc4_connector_node_id {
+	u32 val;
+	struct {
+		u32 vindex:8;
+		u32 dma_type:5;
+		u32 rsvd:19;
+	};
+} __packed __aligned(4);
+
+enum sof_ipc4_probe_runtime_param {
+	SOF_IPC4_PROBE_INJECTION_DMA = 1,
+	SOF_IPC4_PROBE_INJECTION_DMA_DETACH,
+	SOF_IPC4_PROBE_POINTS,
+	SOF_IPC4_PROBE_POINTS_DISCONNECT,
+};
+
+/* For getting probe dmas */
+struct sof_ipc4_probe_dma {
+	union sof_ipc4_connector_node_id node_id;
+	unsigned int dma_buffer_size;
+} __packed __aligned(4);
+
+struct sof_ipc4_probe_gtw_cfg {
+	union sof_ipc4_connector_node_id node_id;
+	u32 dma_buffer_size;
+} __packed __aligned(4);
+
+struct sof_ipc4_probe_cfg {
+	struct sof_ipc4_base_module_cfg base;
+	struct sof_ipc4_probe_gtw_cfg gtw_cfg;
+} __packed __aligned(4);
+
+enum sof_ipc4_probe_type {
+	SOF_IPC4_PROBE_TYPE_INPUT = 0,
+	SOF_IPC4_PROBE_TYPE_OUTPUT,
+	SOF_IPC4_PROBE_TYPE_INTERNAL
+};
+
+union sof_ipc4_probe_point_id {
+	u32 full_id;
+	struct {
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
+		u32 index:6;
```

Comment:
confusing, there's full_id and index, what is the difference or overlap?

Reply:
To be honest I do not know exactly what all these bit-fields mean exactly. The struct was copied from firmware side, and the same struct is apparently used by windows driver too. Somehow the struct members identify the audio buffer to probe. I had to use fw-side prints to find some number that were accepted by the firmware and topology combination. On IPC3 side the buffer_id alone is used for the same purpose.

Reply:
I agree that the naming is confusing (which is not that uncommon within IPC4)...
I would rename the `full_id` to `val` or `val32`. This is used within `sof_ipc4_probe_point` as buffer_id - which is misleading also as it is not really a buffer ID but a composite of assorted parameters describing a buffer...